File: python/Chapter 2/Question2_2.py
from classes.LinkedList import *


# if k = 1, return the last element
def kth_to_last(linkedlist, k):
    if k <= 0:
        return "invalid k"
    pointer2 = linkedlist.head
    for i in range(k-1):
        if pointer2.next != None:
            pointer2 = pointer2.next
        else:
            return "k exceeds the length of linkedlist"
    pointer1 = linkedlist.head
    while pointer2.next != None:
        pointer2 = pointer2.next
        pointer1 = pointer1.next
    return pointer1
   
 

#---------------test------------------
L = randomLinkedList(8, 0, 100)
print L
print "The 3th to last element is", kth_to_last(L, 3)

File: c++/Chapter 2/Question2_2.cpp
#include <iostream>
#include "LinkedListClass/LinkedList.h"
#include "../Chapter 1/Question2_2.h"

// this implementation runs in O(n) time with O(1) space
// the end of the list corresponds to k = 1
Node* ktoLast(Node* head, int k)
{
    Node* current = head;
    int length = 0;

    // find the length of the list
    while (current)
    {
        length++;
        current = current->next;
    }
    
    // reset back to head
    current = head;
    
    if (k > length || k < 1)
    {
        return nullptr;
    }

    int count = 0;
    while (length - count != k)
    {
        count++;
        current = current->next;
    }

    return current;
}

int Question2_2::run()
{
    LinkedList* myList = new LinkedList();
    myList->insert(5);
    myList->insert(7);
    myList->insert(12);
    myList->insert(7);
    myList->insert(16);
    myList->insert(18);
    myList->insert(25);
    myList->insert(11);
    myList->insert(5);

    std::cout << "The list is: ";
    myList->display();

    std::cout << "The 4th to last element is: ";
    std::cout << ktoLast(myList->head, 4)->data << std::endl;
    delete myList;

    return 0;
}